### PR TITLE
Normative: Avoid microtask queue delay when resolved

### PR DESCRIPTION
**Pull request by littledan**
Closes #47, #48, #43

### CODE REVIEW

**Comment by littledan** [comment]
Note, this is one of two possible solutions. I'm working on a writeup of how this would be solved instead through a microtask queue checkpoint.

**Comment by littledan** [comment]
cc @georgneis

**Comment by littledan** [comment]
There is a possible tweak to this patch, where we only treat a module as synchronous if it completed synchronously the first time, and so any top-level await  (even from a previous module graph traversal) will always take a turn through the job queue before the module runs. Maybe this would ameliorate the "Zalgo" concerns a little bit, but it would require extra bookkeeping.

File: spec.html [hunk 1]
```diff
@@ -357,33 +357,40 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
...
           1. Repeat, while _done_ is *false*,
-            1. Let _requiredModule_ be the last element in _stack_.
+            1. Let <del>_requiredModule_<del><ins>_record_</ins> be the last element in _stack_.
+            1. <ins>Let _requiredModule_ be _record_.[[Module]]</ins>
             1. Remove the last element of _stack_.
             1. Set _requiredModule_.[[Status]] to `"evaluated"`.
             1. If _requiredModule_ and _module_ are the same Module Record, set _done_ to *true*.
+            1. <ins>Otherwise, if _module_.[[ExecPromise]] is rejected,</ins>
+              1. <ins>Let _reason_ be _module_.[[ExecPromise]].[[PromiseResult]].</ins>
+              1. <ins>Perform ! Call(_record_.[[Capability]].[[Reject]], *undefined*, &laquo; _reason_ &raquo;).</ins>
```
Comment (littledan) [inline spec.html]: The purpose of this loop is to recover the error setting that used to be in Evaluate(). Namely, if there is a cycle and one thing in the cycle has an error, all modules in the cycle have the error. The error is only propagated in the synchronous case; attempting to propagate in the async case would be converting a resolved Promise to a rejected one, which is not allowed.

Comment (littledan) [reply spec.html]: never mind, this is now subsumed by using the ExecPromise from the root of the cycle.

File: spec.html [hunk 2]
```diff
@@ -357,7 +357,7 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
         1. <ins>Let _evalCapability_ be ! NewPromiseCapability(%Promise%).</ins>
         1. <ins>Set _module_.[[ExecPromise]] to _evalCapability_.[[Promise]].
         1. Set _index_ to _index_ + 1.
-        1. Append _module_ to _stack_.
+        1. Append _module_.
```
Comment (guybedford) [inline spec.html]: This seems like a typo?

Comment (littledan) [reply spec.html]: Fixed

**Comment by littledan** [comment]
Applied https://github.com/tc39/proposal-top-level-await/pull/49#issuecomment-466409808 in  [`ab2344e`](/tc39/proposal-top-level-await/commit/ab2344eb50fa385e32ba2827773ef17ea30de7b6)

**Comment by littledan** [comment]
I would prefer to not take this approach, as it leaves whether a module is async or not up to whether it dynamically hits an await, as opposed to whether one is syntactically present (as in the current proposal). So, closing this PR.

**Pull request closed without merging**